service/controller/app/v1/resource/secret/desired.go

```diff
@@ -35,17 +37,34 @@ func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interf
 		return nil, nil
 	}
 
-	if appSecretName != "" && catalogSecretName != "" {
-		return nil, microerror.Maskf(executionFailedError, "merging app and catalog secrets is not yet supported")
+	// We get the catalog level secrets if configured.
+	catalogData, err := r.getSecretDataForCatalog(ctx, cc.AppCatalog)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// We get the app level secrets if configured.
+	appData, err := r.getSecretDataForApp(ctx, cr)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// Secrets are merged and in case of intersecting values the app level
+	// secrets are preferred.
+	mergedData, err := helmclient.MergeValues(catalogData, appData)
+	if err != nil {
+		return nil, microerror.Mask(err)
 	}
 
-	data, err := r.getSecretData(ctx, cr)
+	secretsYAML, err := yaml.Marshal(mergedData)
 	if err != nil {
 		return nil, microerror.Mask(err)
 	}
 
 	secret := &corev1.Secret{
-		Data: data,
+		Data: map[string][]byte{
+			"values": secretsYAML,
```

Comment:
The key name can have any name. The only restriction in chart-operator is only a single key with YAML is supported.